plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -350,7 +399,29 @@ contract = do
     let theClient = machineClient (scriptInstance sc) sc
     _ <- mapError StateMachineError $ StateMachine.runInitialise theClient (initialState theClient) mempty
     forever $ do
-        mapError RunStepError (endpoint @"run step") >>= mapError StateMachineError . StateMachine.runStep theClient
+        i <- mapError RunStepError (endpoint @"run step")
+        checkTransition theClient sc i
+        mapError StateMachineError $ StateMachine.runStep theClient i
+
+-- | Apply 'checkValidState' to the states before and after a transition
```

Comment:
Should this be something we do in the state machine lib, maybe?

Reply:
It would be nice to do *something* of that kind in the state machine lib. But for `checkValidState` we need the `ConversionRate`, which we only get as part of the `Input`, so I'm not sure how best to wire it through to the validation function in a generic way in the state machine lib.

Reply:
Yeah, I agree it's unclear.